### PR TITLE
Hibernate OGM Infinispan Remote (Hot Rod) Demo

### PR DESCRIPTION
**Pull request by fax4ever**

### CODE REVIEW

**Comment by DavideD** [comment]
Thanks @fax4ever, looks great.

I wonder if we should add some javadoc on top of the integration tests to describe the use case.
Normally I wouldn't, but this is a demo I suspect not everyone is used to this kind of tests.

WDYT?

File: hibernate-ogm/playsport-demo/src/main/java/org/hibernate/ogm/demo/playsport/rest/exception/ExceptionDto.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+/**
+ * Created by fabio on 18/08/2017.
+ */
+package org.hibernate.ogm.demo.playsport.rest.exception;
+
+public class ExceptionDto {
```
Comment (Sanne) [inline hibernate-ogm/playsport-demo/src/main/java/org/hibernate/ogm/demo/playsport/rest/exception/ExceptionDto.java]: Could you add some more javadoc to such helper classes? I'm not using REST very often so I'm not sure what these are for. I guess people studying the demo will want to understand this as well.

File: hibernate-ogm/playsport-demo/src/main/webapp/WEB-INF/beans.xml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<beans xmlns="http://xmlns.jcp.org/xml/ns/javaee" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (Sanne) [inline hibernate-ogm/playsport-demo/src/main/webapp/WEB-INF/beans.xml]: Is this file necessary?

File: hibernate-ogm/playsport-demo/src/test/resources/back_to_5_1_0_Final_ogm_version.patch [hunk 3]
```diff
@@ -0,0 +1,41 @@
+Index: hibernate-ogm/playsport-demo/pom.xml
+IDEA additional info:
```
Comment (Sanne) [inline hibernate-ogm/playsport-demo/src/test/resources/back_to_5_1_0_Final_ogm_version.patch]: What is this file for?

File: hibernate-ogm/playsport-demo/README.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
+# Play Sport: Hibernate OGM (Infinispan Remote) Demo
+
+Hibernate OGM can help write applications that interacts with a nosql database, avoiding writing a lot of "boilerplate code".
+OGM is based on the JPA standard notation and it is integrated with JTA lifecycle.
+ 
+This demo is designed for Wildfly 10 JEE 7 container, but adding some dependencies it can work well as a standalone java application.
+The next version will be tested for Openshift Container Platform.
+
+In this demo, some simplifications are made, such as exposing directly Hibernate entities through Rest services. 
+For more realistic production cases, you can use the framework as mapstruct (see http://mapstruct.org/) to easy translate entities into dtos and vice versa.
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply hibernate-ogm/playsport-demo/README.md]: Yes you right, it is not very related. I take it away.

File: hibernate-ogm/playsport-demo/src/test/resources/back_to_5_1_0_Final_ogm_version.patch (hunk 3 above)
Comment (fax4ever) [reply hibernate-ogm/playsport-demo/src/test/resources/back_to_5_1_0_Final_ogm_version.patch]: It is no longer necessary, I would only use the 5.2.0.Alpha tag. Ok?

File: hibernate-ogm/playsport-demo/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply hibernate-ogm/playsport-demo/README.md]: I actually like the reference to MapStruct ;)

**Comment by gunnarmorling** [comment]
Hi @fax4ever, great work! Very nice to see the new dialect. On the demo, does it focus on different aspects than the existing hike manager demo app (which also is an OGM-based REST application)?

**Comment by fax4ever** [comment]
Hi @gunnarmorling. Thank you. At the moment we only have the REST interface, in the next releases I would like to extend it by adding more interfaces and use cases.

**Comment by gunnarmorling** [comment]
Yeah, my question is is it worth to have (and maintain) two demos which
largely show the same thing or should we rather focus on one and e.g.
extend the existing one if there's new stuff to show off.

2017-09-11 8:32 GMT+02:00 Fabio Massimo <notifications@github.com>:

File: hibernate-ogm/playsport-demo/src/main/java/org/hibernate/ogm/demo/playsport/core/repo/AthleteRepo.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Inject
+    private EntityManager em;
+
+    public void add(@Valid Athlete athlete) {
+        em.persist(athlete);
+    }
+
+    public List<Athlete> findAll() {
+        return em.createQuery("select c from Athlete c", Athlete.class).getResultList();
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply hibernate-ogm/playsport-demo/src/main/java/org/hibernate/ogm/demo/playsport/core/repo/AthleteRepo.java]: It seems that it works well if the indexes are consistent with the entries of the caches. Using hibernate.search.default.directory_provider ram I never had any problems.

[the remark this answers is not in the document]
Comment (Sanne) [reply hibernate-ogm/playsport-demo/src/main/java/org/hibernate/ogm/demo/playsport/core/repo/AthleteRepo.java]: ah right, I forgot that all dialects can use the automatic translation form JPQL to Hibernate Search queries. FYI we still need to implement proper query capabilities for the Hot Rod dialect, so that it doesn't have to maintain local indexes.

File: hibernate-ogm/playsport-demo/src/main/java/org/hibernate/ogm/demo/playsport/core/entity/Club.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+import org.hibernate.annotations.GenericGenerator;
+import org.hibernate.search.annotations.Analyze;
+import org.hibernate.search.annotations.Field;
+import org.hibernate.search.annotations.Indexed;
+import org.hibernate.validator.constraints.NotEmpty;
+
+@Entity
+@Indexed
+@JsonIgnoreProperties({"hibernateLazyInitializer", "handler", "athletes", "employees"})
+public class Club {
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply hibernate-ogm/playsport-demo/src/main/java/org/hibernate/ogm/demo/playsport/core/entity/Club.java]: When a rest service retrive a list of entities from a repositories, the objects are already detached (because transaction boundaries are scoped with service methods), but I have in any case the proxies of objects. On proxy have always the property hibernateLazyInitializer. I probably miss something.

[the remark this answers is not in the document]
Comment (Sanne) [reply hibernate-ogm/playsport-demo/src/main/java/org/hibernate/ogm/demo/playsport/core/entity/Club.java]: Thanks, I didn't know these properties were going to be exposes to the REST API.

File: hibernate-ogm/playsport-demo/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (fax4ever) [reply hibernate-ogm/playsport-demo/README.md]: I've done many projects with MapStruct and I love it so much. Many of our customers use it and are very happy.

**Comment by fax4ever** [comment]
Less code to maintain is always better. If you prefer and find it useful I can bring things done to the hike manager demo app, using the maven profiles we should be able to have a demo that works on both dialects.

**Comment by Sanne** [comment]
We all agree that less code is better but let's also consider that this code exists now. We could simply merge it and agree that demos are only updated as we see fit, it doesn't necessarily imply more maintenance.

Also there's some value in keeping the projects relatively simple. When I had to create a Hot Rod based demo recently it was quite inconvenient that our base infrastructure is able to run so many different dialects, profiles, etc.. it does hinder readability.

**Comment by fax4ever** [comment]
Guys I'm going to close this old PR. I'm going to create a new one, a full working demo for Openshift / RHOAR platforms based on OGM and Infinispan datastore
